infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/replication/fsm/WaitSnapshotApplyState.java

```diff
@@ -0,0 +1,190 @@
+                return this;
+            case SNAPSHOT_APPLY_COMPLETE:
+                UUID snapshotSyncApplyId = event.getMetadata().getRequestId();
+                /*
+                 This is required as in the following sequence of events:
+
+                 1. SNAPSHOT_SYNC_REQUEST (ID = 1) EXTERNAL
+                 2. SYNC_CANCEL (ID = 1) EXTERNAL
+                 3. SNAPSHOT_SYNC_REQUEST (ID = 2) EXTERNAL
+
+                 Snapshot Sync with ID = 1 could be completed in between (1 and 2) but show up in the queue
+                 as 4, attempting to process a completion event for the incorrect snapshot sync.
+                 */
+                if (snapshotSyncApplyId.equals(transitionEventId)) {
+                    LogReplicationState logEntrySyncState = fsm.getStates()
+                            .get(LogReplicationStateType.IN_LOG_ENTRY_SYNC);
+                    // We need to set a new transition event Id, so anything happening on this new state
+                    // is marked with this unique Id and correlated to cancel or trimmed events.
+                    logEntrySyncState.setTransitionEventId(event.getEventID());
+                    fsm.setBaseSnapshot(event.getMetadata().getLastTransferredBaseSnapshot());
+                    fsm.setAckedTimestamp(event.getMetadata().getLastLogEntrySyncedTimestamp());
+                    log.info("Snapshot Sync apply completed, syncRequestId={}, baseSnapshot={}. Transition to LOG_ENTRY_SYNC",
+                            event.getEventID(), event.getMetadata().getLastTransferredBaseSnapshot());
+                    return logEntrySyncState;
+                }
+
+                log.warn("Ignoring snapshot sync apply complete event, for request {}, as ongoing snapshot sync is {}",
+                        snapshotSyncApplyId, transitionEventId);
+                return this;
+            case REPLICATION_STOP:
+                log.debug("Stop Log Replication while waiting for snapshot sync apply to complete id={}", transitionEventId);
+                return fsm.getStates().get(LogReplicationStateType.INITIALIZED);
+            case REPLICATION_SHUTDOWN:
+                log.debug("Shutdown Log Replication while waiting for snapshot sync apply to complete id={}", transitionEventId);
+                return fsm.getStates().get(LogReplicationStateType.STOPPED);
+            default: {
+                log.warn("Unexpected log replication event {} when in wait snapshot sync apply state.", event.getType());
+                throw new IllegalTransitionException(event.getType(), getType());
+            }
+        }
+    }
+
+    @Override
+    public void onEntry(LogReplicationState from) {
+        log.info("OnEntry :: wait snapshot apply state");
+        this.fsm.getLogReplicationFSMWorkers().submit(this::verifyStatusOfSnapshotSyncApply);
+    }
+
+    private void verifyStatusOfSnapshotSyncApply() {
```

Comment:
Looks like we don't have a timeout here, does it mean we will retry to verify forever if standby's apply phase is stuck? In this case, we need a user to enforce snapshot sync?

Reply:
yes, that's the way it will perform.